csharp/ql/src/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -131,340 +178,497 @@ module Public {
+   * parameter node already exists.
+   */
+  private predicate parameterReadState(SummarizedCallable c, SummaryNodeState state, int i) {
+    state.isInputState(c, SummaryComponentStack::argument(i))
+  }
+
+  /**
+   * Holds if a synthesized summary node is needed for the state `state` in summarized
    * callable `c`.
    */
-  predicate internalNodeRange(SummarizedCallable c, SummaryInternalNodeState state) {
-    state.isReadState(c, _, _, _)
+  predicate summaryNodeRange(SummarizedCallable c, SummaryNodeState state) {
+    state.isInputState(c, _) and
+    not parameterReadState(c, state, _)
     or
-    state.isStoreState(c, _, _, _)
+    state.isOutputState(c, _)
   }
 
   pragma[noinline]
-  private Node internalNodeLastReadState(
-    SummarizedCallable c, SummaryInput input, ContentList inputContents
-  ) {
-    exists(SummaryInternalNodeState state |
-      state.isReadState(c, input, inputContents, inputContents.length()) and
-      result = internalNode(c, state)
+  private Node summaryNodeInputState(SummarizedCallable c, SummaryComponentStack s) {
+    exists(SummaryNodeState state | state.isInputState(c, s) |
+      result = summaryNode(c, state)
+      or
+      exists(int i |
+        parameterReadState(c, state, i) and
+        result.(ParameterNode).isParameterOf(c, i)
+      )
     )
   }
 
   pragma[noinline]
-  private Node internalNodeFirstStoreState(
-    SummarizedCallable c, SummaryOutput output, ContentList outputContents
-  ) {
-    exists(SummaryInternalNodeState state |
-      state.isStoreState(c, output, outputContents, outputContents.length()) and
-      result = internalNode(c, state)
+  private Node summaryNodeOutputState(SummarizedCallable c, SummaryComponentStack s) {
+    exists(SummaryNodeState state |
+      state.isOutputState(c, s) and
+      result = summaryNode(c, state)
     )
   }
 
   /**
-   * Holds if there is a local step from `pred` to `succ`, which is synthesized
-   * from a flow summary.
+   * Holds if a write targets `post`, which is a post-update node for the `i`th
+   * parameter of `c`.
+   */
+  private predicate isParameterPostUpdate(Node post, SummarizedCallable c, int i) {
+    post = summaryNodeOutputState(c, SummaryComponentStack::argument(i))
+  }
+
+  /** Holds if a parameter node is required for the `i`th parameter of `c`. */
+  predicate summaryParameterNodeRange(SummarizedCallable c, int i) {
+    parameterReadState(c, _, i)
+    or
+    isParameterPostUpdate(_, c, i)
+  }
+
+  private Node pre(Node post) {
+    summaryPostUpdateNode(post, result)
+    or
+    not summaryPostUpdateNode(post, _) and
+    result = post
+  }
+
+  private predicate callbackOutput(
+    SummarizedCallable c, SummaryComponentStack s, Node receiver, ReturnKind rk
+  ) {
+    any(SummaryNodeState state).isInputState(c, s) and
+    s.head() = TReturnSummaryComponent(rk) and
+    receiver = pre(summaryNodeInputState(c, s.drop(1)))
```

Comment:
What's the reason for the `pre` here?

Reply:
Otherwise flow that passes through a callback inside the summarized callable does not work, as the receiver of the two summaries will be different. Example: `map(f, l)` has two summaries `element of arg 1 -> parameter 0 of arg 0` and `return of arg 0 -> element of return`, and both involve a call to the callback at argument 0. If we do not use `pre`, the first summary will have the post-update node of parameter 0 as the receiver.

Reply:
That's a perfectly good explanation for the use of `pre` in the predicate `callbackInput` below as it pertains to output states. But I'm sure it fully answers the use here?

Reply:
I think it's just always going to be the identity here.

Reply:
Ah, yes you are right; I thought I only added it to `callbackInput`